test/clj/game/cards/upgrades_test.clj

```diff
@@ -3336,3 +3336,243 @@
+    (let [og (get-content state :remote1 0)
+          pad (get-content state :remote1 1)]
+      (rez state :corp og)
+      (rez state :corp pad)
+    (take-credits state :corp)
+    (run-on state "Server 1")
+    (run-continue state)
+    (click-card state :runner pad)
+    (click-prompt state :runner "Pay 7 [Credits] to trash")
+    (is (= "PAD Campaign" (:title (first (:discard (get-corp))))) "PAD Campaign trashed for 7c (3 more than normal cost thanks to Oaktown Grid")
+    (is (= 0 (:credit (get-runner)))))))
+
+(deftest rutherford-grid
+  (do-game
+    (new-game {:corp {:hand ["Rutherford Grid" "Caduceus"]}})
+    (play-from-hand state :corp "Rutherford Grid" "New remote")
+    (play-from-hand state :corp "Caduceus" "Remote 1")
+    (let [rg (get-content state :remote1 0)
+          caduceus (get-ice state :remote1 0)]
+      (rez state :corp rg)
+      (rez state :corp caduceus)
+      (take-credits state :corp)
+      (run-on state "Server 1")
+      (run-continue state)
+      (fire-subs state caduceus)
+      (click-prompt state :corp "3")                        ;;Boost trace by 3
+      (is (= 8 (:strength (get-prompt state :runner))) "3 base, +2 of upgrade +3 boost"))))
+
+(deftest akitaro-watanabe
+  (do-game
+    (new-game {:corp {:hand ["Akitaro Watanabe" "Ice Wall" "Enigma"]}})
+    (play-from-hand state :corp "Akitaro Watanabe" "HQ")
+    (play-from-hand state :corp "Ice Wall" "HQ")
+    (play-from-hand state :corp "Enigma" "HQ")              ;;-1c
+    (let [akitaro (get-content state :hq 0)
+          iw (get-ice state :hq 0)
+          enigma (get-ice state :hq 1)]
+      (rez state :corp akitaro)
+      (rez state :corp iw)
+      (rez state :corp enigma)
+      (is (= 2 (:credit (get-corp))) "2c left after installing and rezzing with AW bonus"))))
+
+(deftest research-station
+  (do-game
+    (new-game {:corp {:id "NBN: The World is Yours*"
+                      :hand [(qty "Research Station" 9)] :deck ["Research Station"]}})
+    (play-from-hand state :corp "Research Station" "HQ")
+    (rez state :corp (get-content state :hq 0))
+    (take-credits state :corp)
+    (is (= 8 (count (:hand (get-corp)))) "+3 cards, 2 of Research Station and +1 of ID")
+    (is (= nil (get-prompt state :corp)) "No prompt asking to discard cards from hand")
+    (take-credits state :runner)
+    (is (= 9 (count (:hand (get-corp)))) "Double check that you start next turn with 8 cards")))
+
+(deftest traffic-analyzer
+  (do-game
+    (new-game {:corp   {:hand ["Traffic Analyzer" "Ice Wall"]}
+               :runner {:tag 2 :hand [(qty "Corroder" 2)]}})
+    (play-from-hand state :corp "Traffic Analyzer" "HQ")
+    (play-from-hand state :corp "Ice Wall" "HQ")
+    (rez state :corp (get-content state :hq 0))
+    (rez state :corp (get-ice state :hq 0))
+    (click-prompt state :corp "4")
+    (click-prompt state :runner "0")
+    (is (= 1 (:credit (get-corp))) "After using all credits to boost trace, Corp gains 1c thanks to Traffic Analyzer")
```

Comment:
This is a good place for `changes-credits` as well. Instead of testing specific amounts (Corp started with X, now they're Y), only test how much has changed (Corp gained 1 when the trace ended).